velox/common/file/FileSystems.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <functional>
+#include <string_view>
+
+#define _VELOX_REGISTER_FILE_SYSTEM(tag) registerFileSystem_##tag
+
+// Registers a filesystem associated with a given tag.
+#define VELOX_REGISTER_FILE_SYSTEM(tag)             \
+  {                                                 \
+    extern void _VELOX_REGISTER_FILE_SYSTEM(tag)(); \
+    _VELOX_REGISTER_FILE_SYSTEM(tag)();             \
+  }
+
+namespace facebook::velox {
+class Config;
+class ReadFile;
+class WriteFile;
+} // namespace facebook::velox
+
+namespace facebook::velox::filesystems {
+
+// An abstract FileSystem
+class FileSystem : public std::enable_shared_from_this<FileSystem> {
+ public:
+  FileSystem(std::shared_ptr<const Config> config) : config_(config) {}
+  virtual ~FileSystem() {}
+  virtual std::string name() const = 0;
+  virtual std::unique_ptr<ReadFile> openReadFile(std::string_view path) = 0;
```

Comment:
nit: perhaps, add a quick comment to each of the methods? Is this API to open file for reading? perhaps, openFileForRead and openFileForWrite or openFile(path, mode) with an enum for mode?

Reply:
I can change to `openFileForRead` and `openFileForWrite`.  `openFile(path, mode)` won't work since the return types are different.

Reply:
> the return types are different.

That's a good point. Thanks for clarifying.